RFC: Policy on semver and API evolution

aturon opened the pull request:
This RFC proposes a comprehensive set of guidelines for which changes to
*stable* APIs are considered breaking from a semver perspective, and which are
not.  These guidelines are intended for both the standard library and for the
crates.io ecosystem.

This does *not* mean that the standard library should be completely free to make
non-semver-breaking changes; there are sometimes still risks of ecosystem pain
that need to be taken into account. Rather, this RFC makes explicit an initial
set of changes that absolutely *cannot* be made without a semver bump.

Along the way, it also discusses some interactions with potential language
features that can help mitigate pain for non-breaking changes.

The RFC covers only API issues; other issues related to language features,
lints, type inference, command line arguments, Cargo, and so on are considered
out of scope.

[Rendered](https://github.com/aturon/rfcs/blob/api-evolution/text/0000-api-evolution.md)

aturon commented:
@bstrie, I know you were interested in this.
@pnkfelix I could use your careful eyes on this one!
cc @alexcrichton @sfackler @brson @huonw @wycats @nikomatsakis

aturon commented:
cc @reem @carllerche -- I think both of you have also expressed some opinions on this topic as well. (This is an important piece of policy, so I want to ensure good visibility).

sfackler commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,748 @@
+  to disambiguate are not automatically "major" changes. (But in such cases, one
+  must evaluate how widespread these "minor" changes are).
+
+* In principle, it should be possible to produce a version of the code for any
```
A slightly weaker way of phrasing this would be "given knowledge of any change, it should be possible to write a version of your code that works both before and after that change". It seems to me like that phrasing is significantly more important to guarantee than the version written, since it means libraries won't be "forced" to abandon old versions of Rust to support new versions.

aturon replied:
Yes, that's also an important constraint, and I'll try to rewrite to clarify.

That said, what's gradually becoming clear to me with this RFC is that we may be able to make a strictly stronger guarantee: that all minor but breaking changes can be worked around automatically with an elaboration process. (I think with a couple of very minor language features, we can get there.) Having a strong guarantee that you don't have to go make changes to your dependencies when you upgrade Rust -- while allowing the standard library to grow -- seems like a great place to be.

sfackler commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,748 @@
+currently private. (Tuple structs with mixtures of public and private fields are
+bad practice in any case.)
+
+#### Minor change: going from a tuple struct with all private fields (with at least one field) to a normal struct, or vice versa.
```
We could also update the compiler to allow both `Foo(..)` and `Foo {..}` if all fields are private. If the type's "totally opaque", it kind of makes sense that the kind of struct would be hidden as well.

reem replied:
I agree with @sfackler, I've viewed this as a wart and we could just fix it.

nikomatsakis replied:
Strictly speaking, `Foo(..)` looks up a match in the value namespace, I believe, though perhaps that is not necessary given our current semantics (basically, for anything where the `Foo(..)` pattern would be legal, it could also check the type namespace now).

huonw commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,748 @@
+
+#### Minor change: adding/removing crate dependencies.
+
+The author is not aware of any possible breakage in altering dependencies (which
```
This may implicitly cause a crate to require nightly if a dep does.

kballard replied:
Perhaps this should just be covered under something like

#### Major change: any change that requires a nightly when the crate did not previously require a nightly

This includes introducing `#[feature]` for the first time, or adding/upgrading a dependency which requires a nightly.

cmr replied:
Additionally, adding/removing dependencies can modify the cargo `feature` set (not `#[feature]`) exposed to upstream, which can implicitly cause breaking changes if you accidentally used a `#[cfg]`d off function.

aturon replied:
@huonw @kballard Thanks, I will take @kballard's suggestion for the wording here.

@cmr That is an interesting scenario, but I don't think it should be considered a major change: if other crates were relying on the feature without specifying it, they're basically breaking the model (and can be easily fixed by opting into the feature themselves).

huonw commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,748 @@
+
+Of course, much of the effect of renaming/moving/removing can be achieved by
+instead using deprecation and `pub use`, and the standard library should not be
+afraid to do so!In the long run, we should consider hiding at least some old
```
(Missing space after !.)

sfackler commented on text/0000-api-evolution.md:
````diff
@@ -0,0 +1,748 @@
+```
+
+because existing uses of `Foo` are shorthand for `Foo<u8>` which yields the
+identical field type.
````
This is not actually the case right now since #213 isn't fully implemented:
```rust
enum Foo<S = i32> {
    Bar,
    Baz(S)
}

fn f<S>(_: Foo<S>) {}

fn main() {
    f(Foo::Bar);
}
```

```
~ ❯ rustc test.rs
test.rs:9:5: 9:6 error: unable to infer enough type information about `_`; type annotations required [E0282]
test.rs:9     f(Foo::Bar);
              ^
error: aborting due to previous error
```

petrochenkov replied:
Yep, until default type parameters drive type inference, generalization can cause surprisingly large breakage. For example [heterogeneous comparisons for `Option`](https://github.com/rust-lang/rfcs/issues/917) can't be implemented now because they would require type annotations on `None` in every comparison and that's a lot of annotations.
Default parameters is an important backward compatibility tool, I hope they'll be finished sooner.

huonw commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,748 @@
+
+For tuple structs, this is only a minor change if furthermore *all* fields are
+currently private. (Tuple structs with mixtures of public and private fields are
+bad practice in any case.)
```
Should we have a lint?

reem commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,748 @@
+See "[Signatures in type definitions](#signatures-in-type-definitions)" for some
+general remarks about changes to the actual types in an `enum` definition.
+
+#### Major change: adding new variants.
```
Does this mean we are in fact freezing `io::ErrorKind`?

sfackler replied:
`ErrorKind` has an unstable variant that prevents stable code from exhaustively matching on it: https://github.com/rust-lang/rust/blob/master/src/libstd/io/error.rs#L118-L123

sfackler commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,748 @@
+considered a breaking change. Crate authors should consider using Cargo
+"features" for their crate to make such use opt-in.
+
+#### Minor change: adding/removing crate dependencies.
```
A somewhat related concept would be changing the major version of a dependency. It seems like that could be either a minor or major change depending on if the crate is visible in your crate's public API.

kballard replied:
Does Cargo allow multiple versions of the same crate to exist in a single dependency tree, or does it restrict each crate to only resolving to a single version? My assumption has been the latter (although some language package managers do the former, such as `npm`). If it's the latter, then any major version upgrade for a dependency could be a breaking change if the dependency already exists anywhere else in the upstream crate's dependency tree.

By the same logic, the addition of any dependency (using anything other than `*` as the version requirement) could be a breaking change.

alexcrichton replied:
Cargo allows multiple versions of a project, but each version must be semver incompatible with all other activated versions. For example `1.0.0` and `2.0.0` are allowed, but `1.0.0` and `1.1.0` is not.

kballard replied:
Interesting. So it sounds like it's still possible for the addition/upgrade of a dependency to be an incompatible change if the version requirement depends on anything other than the major version number, e.g. `foo = ">1.1.0"` would be a breaking change if a different project in the dependency tree had `foo = "=1.0.5"`.

aturon replied:
Hm, this is a very interesting and somewhat thorny point.

My inclination would be to say that it's a minor change to increase the minor version number as long as the spec is of the form `>X.Y.Z`, and basically assume that a fully locked-down spec like `=X.Y.Z` is a corner case that opts out of the API stability guarantees.

Thoughts?

huonw commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,748 @@
+
+According to the basic principles of this RFC, such a change is minor: it is
+always possible to annotate the call `t.foo()` to be more explicit *in advance*
+using UFCS: `<Trait2>::foo(t)`. This kind of annotation could be done
```
This would have to be `Trait2::foo(t)` or `<T as Trait2>::foo(t)`, wouldn't it? (Otherwise it's trying to go via the trait object type?)

huonw commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,748 @@
+All of the changes mentioned below are considered major changes in the context
+of trait methods, since they can break implementors.
+
+#### Major change: adding new arguments.
```
adding/removing

huonw commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,748 @@
+#### Minor change: introducing new lint warnings/errors
+
+Lints are considered advisory, and changes that cause downstream code to receive
+additional lint warnings/errors are still considered "minor" changes.
```
Worth discussing `#[deny(...)]` causing compilation errors in more detail? (cc https://github.com/rust-lang/rfcs/issues/1029 )

aturon replied:
I'm not sure whether this RFC is the right place to tackle that infrastructure, but I'll at least put in a reference.

huonw commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,748 @@
+
+One line of defense against a "minor" change causing significant breakage is the
+nightly release channel: we can get feedback about breakage long before it makes
+even into a beta release.
```
Similarly, the beta releases are designed for exactly this purpose.

kballard commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,748 @@
+#### Major change: adding a private field when all current fields are public.
+
+This change has the effect of making external struct literals impossible to
+write, which can break code irreparably.
```
FWIW, there is a way around this. The obvious answer is to implement `Default` for the struct, allowing literals of the form `Foo { a: 1, b: 2, ..Default::default() }`.

This solution won't work for struct literals found in `const` or `static`s (although a `const` containing a default value for the struct would). But there's plenty of other changes you can make to structs that would break struct literals found in `const` or `static` but not break struct literals found elsewhere (for example, adding a public field that has no public value that can be used in a `const` or `static`). Given that, requiring that any change not break the ability to use struct literals in `const` / `static` might be considered overly restrictive (especially as most types will never find themselves used in that fashion). And if that restriction is not enforced, then it seems plausible to treat a pre-existing implementation of `Default` as allowing for changes that would otherwise break struct literals.

pnkfelix replied:
@kballard I take it you are assuming that [RFC 736](https://github.com/rust-lang/rfcs/blob/master/text/0736-privacy-respecting-fru.md) is reverted in order to enable this pattern?

That is, this pattern *used* to work, but it does not any longer, because the privacy breakage allowed inadvertent exposure of the internal representation and subsequent breakage of global invariants.

kballard replied:
@pnkfelix Oh. I actually had never even seen RFC 736. I missed that change completely.

Incidentally, is there some reason that this was disallowed completely, instead of only for types that aren't `Copy`? Using FRU for a `Copy` type is equivalent to copying it and mutating the public fields. I'll file an RFC on it.

pnkfelix replied:
@kballard (i don't think there was any reason beyond adopting the "simplest" change that could possibly work, where "simplest" is measured according to some hand-wavy ordering w.r.t. the description of the language itself)

Gankro commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,748 @@
+## Behavioral issues
+
+- Is it permitted to change a contract from "abort" to "panic"? What about from
+  "panic" to "return an `Err`"?
```
Exceedingly little code can even *distinguish* between an abort and a panic, so I wouldn't expect this to be a serious problem in practice. The only thing I can see being trouble is unsafe code, which can legitimately rely on the fact that something aborted to avoid setting up guards against unwinding. In such a case this could lead to UB! However if we had no_unwind assertions or something unsafe code could at least guard against such a change. Although adding no_unwind assertions would make this change even-more-breaking!

panic -> Err seems like a tricky semantic change that can cause weird error-handling bugs. For one, if API A declares that it panics on input X, and API B uses API A but *also* transitively reports the panic in one of its APIs for input Y, its documentation will at best be incorrect after an upgrade: it no longer panics on Y. 

Otherwise you may just get bad error handling results from things like "what I thought was an effectively exhaustive match no longer is (new Error kinds are possible -- basically a soft version of adding a variant to an enum)" or "The system reports `frobs are out of sync` on error X since this is the only possibility in 1.0, but since 1.1 it can now also mean `frams are misaligned` (old Error kinds are now more overloaded)". This seems like a major breaking change from a semantic perspective.

oli-obk commented on text/0000-api-evolution.md:
````diff
@@ -0,0 +1,748 @@
+```
+
+This is especially useful for the case where multiple modules being glob
+imported happen to export items with the same name.
````
This could rather be solved by importing neither, but noting the name. When the code uses an item with that name, an error "item not imported due to conflicting glob imports" could list all possibilities and suggest adding an explicit `use`

aturon replied:
Good point! I'll add that proposal as well.

P1start commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,748 @@
+
+#### Major change: adding new fields to a variant.
+
+If the enum is public, so is the full contents of all of its variants. As per
```
s/so is/so are/ – ‘the full contents’ is plural.

P1start commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,748 @@
+At the moment, Rust does not provide defaulted arguments, so any change in arity
+is a breaking change.
+
+#### Minor change: introducing a new type parameter.
```
I believe this goes against [the overview’s ‘principles of the policy’](https://github.com/aturon/rfcs/blob/api-evolution/text/0000-api-evolution.md#principles-of-the-policy):

> That means that any breakage in a minor release must be very "shallow": it must always be possible to locally fix the problem through some kind of disambiguation that could have been done in advance (by using more explicit forms) or other annotation (like disabling a lint). It means that minor changes can never leave you in a state that requires breaking changes to your own code.

As far as I know, there’s not always a way of ‘fix[ing] the problem through some kind of disambiguation that could have been done in advance’ for this kind of change. To explain with an example, if the breaking change example mentioned in the RFC (adding an extra type parameter to `fn foo<T>`) occurred in the standard library between, say, Rust 1.3 and 1.4, and my library (which depended on `std`) had code that called `foo` with an explicit type parameter (like `foo::<u8>()`), I’d have no way of fixing my code in a way that continues to support Rust 1.3 but also supports the newest version of Rust. In other words, I’d have to make a choice between supporting the older versions of Rust that I supported and supporting the newer versions (which have an extra type parameter on `foo`).

The section mentioned above does say ‘in principle’, but it also says that the RFC should suggest ‘additional tooling or language support’, neither of which seems to be outlined in the RFC to address this particular problem.

oli-obk replied:
In almost all situations I can think of, you can probably use type ascription either in the parameters or in the return value. Any situations where this isn't enough probably have global state and are abusing generic code.

P1start replied:
I disagree. To give a more concrete example, imagine that `std` provided a function for calculating the hash of a value:

```rust
pub fn hash<T>(value: &T) -> u64 where T: Hash {
    let mut h = SomeHasher::new();
    value.hash(&mut h);
    h.finish()
}
```

Let’s pretend that I have a library (which depends on `std`) that has code that looks like `hash::<u8>(1)`. Now, in Rust 1.4, the function above has a new type parameter added to make it generic over the hasher used:

```rust
pub fn hash<T, H>(value: &T) -> u64 where H: Default + Hasher, T: Hash {
    let mut h: H = Default::default();
    value.hash(&mut h);
    h.finish()
}
```

According to the RFC, this would be a minor change. However, there is no way for me to update my code in a way that supports both 1.3 and 1.4—in 1.4, I *have* to specify the second type parameter, but in 1.3, I can’t.

I suppose this doesn’t apply to most functions (only those where not all type parameters can be inferred), but I think that the RFC should at least specify that this is only a minor change if there is a way of inferring the type of all type parameters (new and old).

kballard replied:
@P1start Your example there *can* infer its type parameter, and your example invocation `hash::<u8>(1)` won't work to begin with.

huonw replied:
I wonder if we could allow hinting only a prefix of the generic type parameters, this would make situations like this where the new type parameter is inferrable always just work. I have at least one API where this would be useful myself (functions are usually called like either `foo::<X, _>()`, with the `_` essentially never needing to be specified).

P1start replied:
@kballard The second type parameter doesn’t seem to be inferred [when using `std`’s existing `Hash[er]` traits](http://is.gd/aYVKWt). And the invocation should have been `hash::<u8>(&1)` (but that’s just a mistake and is beside the point).

kballard replied:
@P1start I was referring to that first code snippet. With the second code snippet, the addition of a non-inferrable type parameter would of course be a breaking change, and I would expect anyone making such a change to provide a default value for the type parameter (or if they really don't want to, then to consider it a major version bump).

nikomatsakis commented:
@aturon I'll go over in detail, but one thought I had is that it might be useful to elaborate a kind of "best practices for forwards compatibility". That's probably not part of this RFC, but it might help illuminate nonetheless.

nagisa commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,748 @@
+Along the way, it also discusses some interactions with potential language
+features that can help mitigate pain for non-breaking changes.
+
+The RFC covers only API issues; other issues related to language features,
```
Redundant paragraph. Language, lints, type inference, CLI are not part of standard library (first paragraph), nor they are a part of crates.io ecosystem.

pnkfelix replied:
In the past, for better or worse, sometimes changes to things like command line arguments to `rustc` were for various reasoned classified as "library issues" (as opposed to language issues).

While the first paragraph does indeed restrict the domain to "standard library" and "crates.io ecosystem", I think this paragraph can stay -- perhaps just modify it to make it clear that this paragraph is in fact a consequence a fact that all these things are not part of the standard library nor crates.io ecosystem.

SimonSapin commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,748 @@
+  "panic" to "return an `Err`"?
+
+- Should we try to lay out more specific guidance for behavioral changes at this
+  point?
```
There are behavior changes like [#23951 Changing the meaning of the count to splitn](https://github.com/rust-lang/rust/pull/23951) that are not *breaking* per the definition in this RFC (they do not cause downstream code to fail to compile), but still should be considered *major* as they break basic expectations that downstream code could reasonably make.

Unfortunately I don’t see a clear criteria to distinguish those from behavior changes that are more reasonable or expected like changing results of `char::is_alphabetic` when updating to a new Unicode version.

aturon replied:
@SimonSapin Did you see the earlier section on behavioral changes? The rough guideline would be that changes that break the explicit, documented behavioral contract of a function are considered breaking.

BurntSushi commented:
I very much love this RFC. Nice work @aturon!

Havvy commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,748 @@
+Adding any item without a default will immediately break all trait implementations.
+
+It's possible that in the future we will allow some kind of
+"[sealing](#sealed-traits)" to say that a trait can only be used as a bound, not
```
This link is dead.

theemathas commented:
What happens if someone finds *another* way to cause unsafty using safe code and the standard library? We already had `thread::scoped` and `Vec::drain`.

tbu- commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,748 @@
+
+As with "[Signatures in type definitions](#signatures-in-type-definitions)",
+traits are permitted to add new type parameters as long as defaults are provided
+(which is backwards compatible).
```
This is not true, consider the change to `Iterator::sum`, which (even though it specifies a default for the type parameter) can't be inferred everywhere.

huonw replied:
I suspect that is the same problem discussed below: https://github.com/rust-lang/rfcs/pull/1105/files#r29635875

apasel422 commented:
Should this mention changing an object-safe trait in such a way that it is no longer object-safe?

theemathas commented:
What about changing an unsafe function to a safe function?

ruud-v-a commented:
There are two angles to versioning, and I think the difference is important:

1. **Given that the next release will not be major, can we include this change?**
2. **Given that we are going to include this change, what should the version number for the next release be?**

If I understand the release train model correctly, the version number is determined before changes that will be included in that version are made, so Rust is in scenario 1. If major bumps are going to be uncommon, then it makes sense to downplay some technically breaking changes as minor.

For Crates.io, I expect the vast majority of crates to be in scenario 2. I think the pragmatic approach is a good one; adding a public item without changing anything else certainly _feels_ like a new feature, not like a backwards incompatible change. **However, to play the devil’s advocate**: there appears to be a hidden assumption that a major version bump is bad and should be avoided. I agree that backwards incompatible changes should be avoided if possible, but this RFC is about _naming only_. If a change is technically breaking and will be included in the next release, the only question is what the version number should be. Why downplay technically breaking changes as minor? To quote the sidebar rules:

> 5 . Chill out! A ~~programming language~~ version number is a silly thing to get upset over.

kballard commented:
> Why downplay technically breaking changes as minor?

Because unnecessarily incrementing the major version number *is* bad. It requires all dependent crates to update their `Cargo.toml` or they won't get any of the bug fixes in this (or future) versions. And it can cause larger dependency trees to end up with multiple major version numbers of the same crate, which always has the potential to behave badly (for example, two crates that both work with types from a third crate, if they're using different versions of the third crate then the types they're working with aren't compatible; or alternatively, a crate that manages some static data, two versions of the crates means two separate instances of the static data). And of course there's the simpler reason which is that any time there's a major version bump, the authors of any dependent crates are going to have to do research to find out what backwards-incompatible changes were made, which is a burden we shouldn't put on developers unnecessarily.

glaebhoerl commented:
I think this seems like the right policy, pragmatically speaking. Essentially, to re-summarize: patch-level version bumps are bugfixes only, backwards *and forwards* compatible; minor versions can cause minor, rare, *non-cascading* and *backwards-compatibly fixable*<sup>1</sup> compile failures due to additions or generalizations; and major versions are for everything else.

<sup>1</sup> These two are very important! At the outset, the idea that "it turns out you can't actually change anything in Rust without the potential for breaking downstreams, but admitting that would be too awkward, so let's just ignore it and pretend" had me rather skeptical, but these are actually very substantial distinctions and guarantees that I'm not sure I could have thought of myself, and they have me convinced that this is the right resolution. So, good work.

The 'elaboration' idea is very clever. I think it can be taken even further:

 * We could also use it to detect the cases where code continues to compile, but a name/method resolves to a different entity than it did before (which I think is *much worse* than if the code had just stopped compiling - guarding against this kind of thing is what a type system is supposed to be for!). If we have the elaborated source from the previous version, we can just check whether it's consistent with the new one.

 * Similarly, we could use this to *automatically fix* the kind of "minor breakage" that a minor version can introduce. If, e.g., the resolution of a name is ambiguous with the new version, just replace it with the fully-resolved name from the previous elaborated source.

(I'm not saying this would be *easy* - but it seems *possible*. And possibly worthwhile.)

pnkfelix commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,748 @@
+  must evaluate how widespread these "minor" changes are).
+
+* In principle, it should be possible to produce a version of the code for any
+  dependencies that *will not break* when upgrading to a new minor
```
@aturon you already said you're planning to rephrase this text, but I just wanted to point out: the wording here is awkward, because the word "that" in "that will not break" is ambiguous as to whether it refers to "the code" or the "any dependencies".

pnkfelix commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,748 @@
+
+Changing a crate from working on stable Rust to *requiring* a nightly is
+considered a breaking change. Crate authors should consider using Cargo
+"features" for their crate to make such use opt-in.
```
can you provide a pointer to the relevant cargo documentation on these `Cargo "features"` ?

Gankro replied:
http://doc.crates.io/manifest.html#the-%5Bfeatures%5D-section

To be clear: this is not that there are some vague set of features that cargo itself provides, but rather that crates can register features of itself to be optionally compiled.

aturon replied:
Ha yes, sorry about that ambiguity. Thanks for the link, @Gankro

pnkfelix commented on text/0000-api-evolution.md:
````diff
@@ -0,0 +1,748 @@
+```
+
+The problem here is that glob imports currently do not allow any of their
+imports to be shadowed by an explicitly-define item.
````
typo: "explicitly-defined"

pnkfelix commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,748 @@
+   implemented. The details are left to a future RFC, however.
+
+2. Even if that change were made, though, there is still the case where two glob
+   imports conflict without any explicit definition "covering" them. This is
```
nit: should say, for precision, "two glob imports conflict *with each other*"

pnkfelix commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,748 @@
+This change retains the ability to use struct literals, but it breaks existing
+uses of such literals; it likewise breaks exhaustive matches against the struct.
+
+#### Minor change: adding or removing private fields when at least one already exists.
```
nit: at least one already exists prior to the change *and also after the change*.

(that is, as written it sounds like removing the sole private field is a minor change, when it is in fact a major change.)

pnkfelix replied:
also, why is it a minor change?  Because it breaks ABI?  (It doesn't seem to qualify as a feature addition on its own; I can imagine bug-fixes that work via an added private field...)

aturon replied:
> also, why is it a minor change?  Because it breaks ABI?  (It doesn't seem to qualify as a feature addition on its own; I can imagine bug-fixes that work via an added private field...)

The labeling of "minor" here is in opposition to major, but is also implicitly allowed in a patch release.

pnkfelix commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,748 @@
+clients of the trait, but would break existing implementors. (Note, as above,
+the potential for "sealed" traits to alter this dynamic.)
+
+#### Minor change: adding a defaulted item.
```
Adding a defaulted method to a trait that has any associated types with a default should probably be considered a major change.

In particular, due to [Default handling in RFC 195](https://github.com/rust-lang/rfcs/blob/master/text/0195-associated-items.md#defaults), if I have an impl that overrides a defaulted associated type, then I will be obligated to now implement the associated method that the upstream code added, regardless of the fact that it carried a default.  (At least, that's my memory of how that rule is to be interpreted.)

pnkfelix replied:
Likewise, ... doesn't this also mean that adding a default type expression to an associated type that did not have one previously should *also* be considered a major change?

pnkfelix commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,748 @@
+runtime behavior. The case where the same method for the same type has
+meaningfully different behavior is considered unlikely enough that the RFC is
+willing to permit it to be labeled as a minor change -- and otherwise, inherent
+methods could never be added after the fact.
```
hmm ... I wonder if this could provide motivation for some way to `use` a type without bringing its inherent impls into scope.

pnkfelix commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,748 @@
+
+Loosening bounds, on the other hand, cannot break code because when you
+reference `Foo<A>`, you *do not learn anything about the bounds on `A`*. (This
+is why you have to repeat any relevant bounds in `impl` blocks for `Foo`, for
```
are we committed to that semantics for `impl` blocks for the long term?  At times I thought I've heard @nikomatsakis muse about various strategies to allow some inference to occur in such cases.

aturon replied:
It's not completely clear, but this RFC can be revised if or when [implied bounds](http://smallcultfollowing.com/babysteps/blog/2014/07/06/implied-bounds/) happen.

pnkfelix commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,748 @@
+#### Minor change: generalizing to generics.
+
+A struct or enum field can change from a concrete type to a generic type
+parameter, provided that the change results in an identical type for all
```
nit: "to a generic type parameter with a default"

pnkfelix commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,748 @@
+# Unresolved questions
+
+## Behavioral issues
+
```
Reading the first item below, I wondered: what does adding a `#[must_use]` annotation to a type qualify as?  Seems like a major change to me -- or is that lint only warn-by-default?

And are there other attributes in this bucket, where adding them should be considered a major change?  (E.g. changes to `#[inline]` directives are probably safe, but then there's stuff like `#[unsafe_no_drop_flag]` ... though  I guess that is unstable anyway, right?)

sfackler replied:
must_use is warn by default.

theemathas commented on text/0000-api-evolution.md:
````diff
@@ -0,0 +1,748 @@
+```
+
+(The use of `?Sized` is essential; otherwise you couldn't recover the original
+signature).
````
Is it OK to go the other way around?

```rust
// Before
fn foo<T: Trait + ?Sized>(x: Box<T>);

// After
fn foo(x: Box<Trait>);

reem replied:
No, since that is less general and can break users in non-easily-fixable ways.

theemathas replied:
@reem Can you give an example?

huonw replied:
I suspect coercions save the day for that sort of example, but more complicated examples with nested types where coercions don't/can't work will certainly break:

```rust
fn foo<T: Trait + ?Sized>(x: &[Box<T>]) {}
```

can be called with `&[Box<i8>]` (if `i8: Trait`) but the `&[Box<Trait>]` version cannot.

Kimundi replied:
@theemathas: You can not create trait objects for unsized types, so if eg `str` implemented `Trait` you could put it in the "before" case, but not in the "after" case.

pnkfelix commented:
One subject that the RFC should probably address but currently does not: How to classify changes to lifetime parameters and lifetime bounds ?

E.g. it seems like this rewrite is a patch-change (not even minor):
```rust
fn foo(&self) -> &int; <==> fn foo<'a>(&'a self) -> &'a int;
```

There is also this kind of rewrite:
```rust
fn foo<'a>(x: &'a Trait); <==> fn foo<'a>(x: &'a (Trait +'a));
```

and then there is this one, the validity of which may change based on [ongoing discussion](https://github.com/rust-lang/rust/pull/25212#issuecomment-100441853):

```rust
fn foo<'a>(x: &'a Box<Trait>); <==> fn foo<'a>(x: &'a Box<Trait+'a>);
```

aturon commented:
Thanks everyone for the feedback! I've pushed a slew of updates to the RFC addressing most of the points that have been raised. I'll also respond to a few general comments below, and do another round of revisions soon.

====

@nikomatsakis 

> I'll go over in detail, but one thought I had is that it might be useful to elaborate a kind of "best practices for forwards compatibility". That's probably not part of this RFC, but it might help illuminate nonetheless.

That sounds interesting, but I'm not entirely sure what you have in mind. Could you give a couple examples?

====

@BurntSushi 

> I very much love this RFC. Nice work @aturon!

Thanks!!

====

@theemathas 

> What happens if someone finds *another* way to cause unsafty using safe code and the standard library? We already had `thread::scoped` and `Vec::drain`.

@nikomatsakis is working on a separate RFC to talk about soundness issues, where the policy is somewhat different than in this RFC.

====

@apasel422 

> Should this mention changing an object-safe trait in such a way that it is no longer object-safe?

Good point! I'm adding text to that effect.

====

@theemathas 

> What about changing an unsafe function to a safe function? 

Good question. Right now, this is a breaking change due to lack of subtyping. I'm not 100% sure what the right policy is here.

====

@glaebhoerl 

I agree completely with your analysis. This RFC isn't digging too deeply into the specifics of elaboration, but tooling along the lines you're suggesting sounds very interesting indeed.

====

@pnkfelix 

I've done an initial round addressing some of your comments, but need to revisit others. I'll let you know when I think I've addressed everything.

pnkfelix commented:
Changes within macros are also a potential source of code breakage.

While the language strived to make `macro_rules!` hygienic, there are places where it simply it is not.  (Some of these cases might get fixed in the future, I am not sure how many we can fix in a backwards-compatible fashion, though.)

For example:
 * https://github.com/rust-lang/rust/issues/22462
 * https://github.com/rust-lang/rust/issues/25345

My point in listing the above examples is that one a crate or the stdlib provides a `macro_rules!` style macro, as long as the bugs above persist, then one must be very careful in what changes one might make to the content of that macro.  Transformations like supposed "alpha-renaming" may not actually apply.

bstrie commented:
This is looking great! An incredible effort. :heart:

bstrie commented:
Would it be worthwhile to go through and study the regression reports from the past four weeks and characterize each occurrence of breakage according to the categories presented here?

bstrie commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,798 @@
+always be possible to locally fix the problem through some kind of
+disambiguation *that could have been done in advance* (by using more explicit
+forms) or other annotation (like disabling a lint). It means that minor changes
+can never leave you in a state that requires breaking changes to your own code.
```
The "could have been done in advance" rule is a lovely formulation. It makes communicating the scope of potential breakage easy and eases the fear of wanton flaunting of semver. Well done!

bstrie commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,798 @@
+problems.
+
+Any breaking, but minor change to the standard library must be evaluated through
+Crater before being committed.
```
I don't think we want to commit to running Crater with *every* minor change. Is the current cadence of once-a-week reports too slow?

bstrie commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,798 @@
+public item can break downstream code arbitrarily.
+
+It would be much better for API evolution (and for ergonomics and intuition) if
+explicitly-defined items trump glob imports. But this is left to a future RFC.
```
Is there any possibility that the changes to name resolution from such an RFC would be a breaking change? Are we allowing ourselves that possibility because "we intended it to be this way"?

bluss commented:
What about changing a crate from depending on Rust 1.0 to Rust 1.1 (and so on)? This is a non-breaking change I assume.

bluss commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,798 @@
+
+One hazard is that with today's explicit type parameter syntax, you must always
+specify *all* type parameters: `foo::<T, U>(x, y)`. That means that adding a new
+type parameter to `foo` can break code, even if a default is provided.
```
default parameters for functions & methods aren't fully implemented, they don't seem to work really.

Stebalien commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,798 @@
+
+This change retains the ability to use struct literals, but it breaks existing
+uses of such literals; it likewise breaks exhaustive matches against the struct.
+
```
I'd explicitly exclude the following design pattern from this restriction:
```rust
pub struct MyOptions {
    pub option_1: usize,
    pub opiton_2: &str,
    #[doc(hidden)]
    pub _non_exhaustive: (),
}
pub enum SomeError {
    A, B, C,
    #[doc(hidden)]
    _non_exhaustive,
}
```

Stebalien commented on text/0000-api-evolution.md:
```diff
@@ -0,0 +1,798 @@
+abstracting a concrete method to use generics instead might work fine for
+clients of the trait, but would break existing implementors. (Note, as above,
+the potential for "sealed" traits to alter this dynamic.)
+
```
I believe removing a `Self: Sized` constraint from non-generic (i.e. otherwise object safe) functions can't break anything. Argument:

1. `impl<T> Trait for T where T: ...` isn't possible outside of the crate that defines `Trait` so we never have to deal with the case where Self could be either Sized or unsized.
2. Self is sized: Any explicit `Self: Sized` bounds on the function are redundant.
3. Self is not sized: The function couldn't have been implemented before the `Self: Sized` bound was removed.

This argument does not generalize to `T: Sized` where T is an parameter type because 1 doesn't apply in this case.

See: rust-lang/rust#25471

Stebalien commented:
It would be kind of nice to include a machine readable API change file in minor updates. Cargo could use this to at least prevent silent breakage. Theoretically, cargo could also use this to automatically disambiguate but I don't know if we really want to rewrite user code and take responsibility for any bugs that might result (we'd have to be REALLY careful and would probably want to prompt the user for every change).

aturon commented:
Sorry for the delayed responses, I'm digging out from a bunch of stuff near the 1.0 release.

One general question that @wycats has raised offline with me is: do we need to require a full implementation of elaboration before we can fully embrace the policy here? Certainly this RFC is strengthened by various implementation work (glob import refinements probably being the most important), but my feeling is that we can adopt this basic framework even without working elaboration, but be especially judicious with checking changes against Crater for breakage. But I'd be curious to hear others' thoughts on this.

====

@pnkfelix 

Regarding macros: this is definitely something we should think about, but I'm much less confident about what to say there at the moment, and we are also relatively conservative about the macros we export from `std`. I'm inclined to punt on these questions for now.

====

@bstrie 

> Would it be worthwhile to go through and study the regression reports from the past four weeks and characterize each occurrence of breakage according to the categories presented here?

Hm, I don't think so -- most of the breakage in the runup to 1.0 would fall into the "requires major version bump", I believe.

> I don't think we want to commit to running Crater with *every* minor change. Is the current cadence of once-a-week reports too slow?

That's a good point. It may be worth further specifying which minor changes need a look on Crater. In particular, simply adding a public item shouldn't require it, even though it could break glob imports -- partly because we should make globs more robust, but partly because we wouldn't be deterred from making a change for that reason (I don't think). Whereas adding a defaulted method to a trait should *definitely* be tested on Crater for conflicts.

Regarding the existing regression reports: I was imagining this happening during prototyping/early implementation, rather than after landing in nightly. But maybe it's enough to land PRs directly to nightly as assess the breakage there, and then consider yanking breaking changes well before stabilization.

One other thing to keep in mind: a stable build still ships with all of the nightly APIs; it just lints on them. That means that basically all breaking changes would be immediately apparent on stable builds too; there's not a lot of margin for error here.

> Is there any possibility that the changes to name resolution from such an RFC would be a breaking change? Are we allowing ourselves that possibility because "we intended it to be this way"?

Yes, it's possible, though @nrc took some pains to future-proof resolution for exactly this reason. And yes, I think these can largely be considered "bug fixes". But we should try to do them soon.

====

@bluss 

> What about changing a crate from depending on Rust 1.0 to Rust 1.1 (and so on)? This is a non-breaking change I assume.

Yes, I would agree -- in particular, if we assume that upgrading from 1.0 to 1.1 is itself not breaking (in a major way), then requiring that upgrade shouldn't count either.

> default type parameters for functions & methods aren't fully implemented, they don't seem to work really. So this can be addressed partially by just finishing that.

Yep, we definitely need some implementation work in this area to make this RFC fully fly.

====

@Stebalien 

> I'd explicitly exclude the following design pattern from this restriction ...

Seems fair, but even better would be to land the equivalent first-class language features. :)

comex commented:
An idea:

The original version of #1122 proposed having a Rust version attribute in Cargo.toml.  That has since been removed in favor of a more minimal PR, but suppose it ends up happening.  What if rustc automatically versioned the standard library based on that number, preventing minor changes from actually breaking anything (in a better way than elaborated source, see below)?  All `#![stable]` declarations in the standard library already have an associated Rust version number; this wouldn't be quite sufficient to take care of all ambiguities, but it'd be close.  Specifically:

- When resolving dispatch ambiguity for a method call, if some of the possible methods are stable in a version later than the caller crate's (or unstable), and others aren't, discard the former.
- Ditto for glob imports: while having explicit items shadow glob-imported ones avoids problems in most use cases, there is still the case of conflict between multiple glob imports, which can be resolved in the same way.
- Optionally: rather than using version comparison to resolve conflict, simply forbid newer items from being used by older crates altogether, in the same way as unstable items.
- New template parameters: add a separate stability attribute for template parameters and allow older crates to omit them without `...`.  Or just allow that in all cases, might make things simpler.
- Introducing generics: this is the ugliest one, but compatibility can be solved by having a mechanism for the standard library to specify the old version of the signature, then having type inference use that version for older code, while still checking afterward that the resolved parameters match the new signature.

The last two sound a bit ugly, but I think it should be fairly rare for such changes to existing items to be made in the first place.  Implementation of the type inference change sounds like it could be a challenge.

So why go to this effort, if all minor changes are designed to break little code in the first place?  A few reasons:

- A `rustfix` that was fully automated would have to do the same work (including the ugly bits with type inference), unless it resorted to full elaboration (i.e. sticking types in everywhere whether needed or not), which would work as a quick fix but be pretty useless as an upgrade path.  Intermediate solutions may be possible, though.

- Elaborated source would require either having an already built copy of your library dependency or downloading a specific old version of rustc.  This means that while cargo could do the conversion automatically, it would be a major usability hassle.  In contrast, by maintaining the required information in the current source as above, users would not suffer if libraries just sat on older versions forever - which sounds gross, but is going to happen one way or another, because code often gets abandoned by its authors.

- With such a robust upgrade path, it would be possible to be *slightly* more liberal than this RFC proposes in making small breaking changes.  Of course, there are many huge downsides to breaking things, period - examples on the Internet, upgrade hassle, etc. - and I don't have any specific suggestion about what sorts of changes could be allowed - but especially combined with Crater, there is at least a bit more wiggle room.

- And if it turns out to be desirable to release a Rust 2.0 in 1 year or 10, the same mechanisms could be used to avoid a Python 3-like flag day.  (Language changes would probably make this trickier than usual, but that's a matter for the other RFC - it's my opinion that enough languages have successfully pulled this off that Rust should be able to, too.)

alexcrichton commented:
This RFC is now entering the week-long final-comment period.

huonw commented:
I'm in favour, but I think that we will *definitely* want to be "quick" to make amendments (to the process, even if we don't literally change the text here) in future as we gain more experience, and as the ecosystem/community changes.

(I'm also a little nervous about https://github.com/rust-lang/rfcs/pull/1105#discussion_r29658339 .)

alexcrichton commented:
The consensus of the library subteam is to merge this RFC. We may tweak the specifics here and there over time, but there is broad consensus among the core ideas behind this RFC and minor updates are always fine to make later!

aturon commented:
It occurred to me today that this RFC did not take into account the impact of "OIBIT" traits (which have a .. impl). In particular, these traits can introduce downstream sensitivity to every aspect of a data type's representation, even if that representation is private.

Effectively, OIBITs make it possible for downstream crates to make promises on behalf of upstream crates that can easily be broken.

I don't think this should change anything about the policy itself, but it'd be worth amending the RFC to discuss it.

[RFC issue](https://github.com/rust-lang/rfcs/issues/1295)
